### PR TITLE
Generic receiver

### PR DESCRIPTION
**Pull request by John-Wiens**
This PR adds a Generic UDP receiver that can receive the following message types: MAP, SPaT, BSM, TIM, SRM, SSM, TIM. 

The generic receiver allows the ODE to receive multiple different message types using a single UDP port. A new UDP port allocation 44990 has been created to act as the generic receiver port. 

Generic Message decoding is done by first identifying the message type, than using existing logic to decode the message and send it to the relevant decoder topic. This identification strategy has been in use by Maricopa County since November 2023. This PR formalizes the strategy developed for Maricopa County and integrates it with the new receiver model developed in March and May 2024. 

Summary of Changes
- The docker-compose file has been updated to open port 44990
- The SVCS Ode Properties file has been updated to include buffer size and port information for the new receiver
- A new Generic Receiver has been created similar to existing receivers for other message types
- UperUtils has been updated to update the strategy for determine message types
- Additional Message Types have been added to the available message types in UperUtils
- Logic to build Json from a Hex packet has been removed from individual receiver classes and consolidated in a single UdpHexDecoder class. The purpose of this change is to allow the generic receiver and existing receivers to call the same code for decoding a given message type. This will remove the need to maintain the logic in multiple locations
- A new test UDP sender has been added which sends all of the different message types to the generic receiver endpoint for testing.

## Related Issue

Related Issue: https://github.com/usdot-jpo-ode/jpo-ode/issues/553

## Motivation and Context

The purpose of this update is to allow support for RSU units that are able to forward messages via UDP, but unable to forward different messages to different ports. 

## How Has This Been Tested?

I have created a new python udp sender script (udpsender_generic.py) which sends the following message types to port 44990: MAP, SPaT, BSM, TIM, SRM, SSM, TIM. For each message type, I monitored the corresponding decoded message topic (OdeMapJson, OdeSpatJson, etc) to ensure messages were decoded properly on the output topic. The generic receiver test script also sends a bad invalid message to the receiver to ensure the system handles invalid messages safely. 

As this change refactors the existing udp receivers, I have also validated that the existing udpsender scripts function as expected and create decoded output on the corresponding decoder topics.

## Types of changes

- [ ] Defect fix (non-breaking change that fixes an issue)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that cause existing functionality to change)

## Checklist:

- [x] I have added any new packages to the sonar-scanner.properties file
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
[ODE Contributing Guide](https://github.com/usdot-jpo-ode/jpo-ode/blob/bugfix/Pull_request_template/docs/contributing_guide.md) 
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: docker-compose.yml [hunk 1]
```diff
@@ -48,6 +48,7 @@ services:
       - "44920:44920/udp"
       - "44930:44930/udp"
       - "44940:44940/udp"
+      - "44990:44990/udp"
```
Comment (mwodahl) [inline docker-compose.yml:51]: Does this port need to be exposed on any of the other docker-compose files?

**Review by mwodahl — COMMENTED** [review]
This looks good and worked well for me when run locally - I just had two small questions.

File: jpo-ode-svcs/src/main/java/us/dot/its/jpo/ode/udp/UdpHexDecoder.java [hunk 2]
```diff
@@ -0,0 +1,245 @@
...
+      // convert bytes to hex string and verify identity
+      String payloadHexString = HexUtils.toHexString(payload).toLowerCase();
+      if (payloadHexString.indexOf(startFlag) == -1)
+         return null;
+      
+      logger.debug("Full {} packet: {}", msgType, payloadHexString);
+      payloadHexString = UperUtil.stripDot3Header(payloadHexString, startFlag);
+      logger.debug("Stripped {} packet: {}", msgType, payloadHexString);
+
+      OdeAsn1Payload timPayload = new OdeAsn1Payload(HexUtils.fromHexString(payloadHexString));
```
[the remark this answers is not in the document]
Comment (John-Wiens) [reply jpo-ode-svcs/src/main/java/us/dot/its/jpo/ode/udp/UdpHexDecoder.java]: Yes. Seems like a good idea to me. Fixed in latest update.

File: docker-compose.yml (hunk 1 above)
Comment (John-Wiens) [reply docker-compose.yml:51]: It looks like this port should be opened in the confluent version of the docker-compose as well. Fixed in the latest update.

**Review by dmccoystephenson — APPROVED** [review]
This looks good to me!

File: jpo-ode-svcs/src/main/java/us/dot/its/jpo/ode/udp/UdpHexDecoder.java [hunk 3]
```diff
@@ -0,0 +1,245 @@
...
+         return null;
+      OdeTimMetadata timMetadata = new OdeTimMetadata(timPayload);
+
+      // Add header data for the decoding process
+      ZonedDateTime utc = ZonedDateTime.now(ZoneOffset.UTC);
+      String timestamp = utc.format(DateTimeFormatter.ofPattern("yyyy-MM-dd'T'HH:mm:ss.SSS'Z'"));
+      timMetadata.setOdeReceivedAt(timestamp);
+
+      timMetadata.setOriginIp(senderIp);
+      timMetadata.setRecordType(RecordType.timMsg);
```
Comment (dan-du-car) [inline jpo-ode-svcs/src/main/java/us/dot/its/jpo/ode/udp/UdpHexDecoder.java:129]: Do we miss a TimSource here?

File: jpo-ode-svcs/src/main/java/us/dot/its/jpo/ode/udp/UdpHexDecoder.java [hunk 4]
```diff
@@ -0,0 +1,245 @@
...
+        int senderPort = packet.getPort();
+        logger.debug("Packet received from {}:{}", senderIp, senderPort);
+
+        // Create OdeMsgPayload and OdeLogMetadata objects and populate them
+        OdeAsn1Payload psmPayload = getPayloadHexString(packet, UperUtil.SupportedMessageTypes.PSM);
+        if (psmPayload == null)
+            return null;
+        OdePsmMetadata psmMetadata = new OdePsmMetadata(psmPayload);
+        // Add header data for the decoding process
+        ZonedDateTime utc = ZonedDateTime.now(ZoneOffset.UTC);
```
Comment (dan-du-car) [inline jpo-ode-svcs/src/main/java/us/dot/its/jpo/ode/udp/UdpHexDecoder.java:231]: We could probably use a method for creating this timestamp if duplicated in multiple places.

File: jpo-ode-svcs/src/main/java/us/dot/its/jpo/ode/udp/bsm/BsmReceiver.java [hunk 5]
```diff
@@ -58,45 +43,21 @@ public void run() {
...
-               bsmMetadata.setRecordGeneratedBy(GeneratedBy.OBU);
-               bsmMetadata.setSecurityResultCode(SecurityResultCode.success);
-
-               // Submit JSON to the OdeRawEncodedMessageJson Kafka Topic
-               bsmPublisher.publish(JsonUtils.toJson(new OdeAsn1Data(bsmMetadata, bsmPayload), false),
-                  bsmPublisher.getOdeProperties().getKafkaTopicOdeRawEncodedBSMJson());
+               String bsmJson = UdpHexDecoder.buildJsonBsmFromPacket(packet);
+
+               if(bsmJson != null){
+                  // Submit JSON to the OdeRawEncodedMessageJson Kafka Topic
```
Comment (dan-du-car) [inline jpo-ode-svcs/src/main/java/us/dot/its/jpo/ode/udp/bsm/BsmReceiver.java:50]: The comment shall be OdeRawEncoded`BSM`Json?

File: jpo-ode-svcs/src/main/java/us/dot/its/jpo/ode/udp/controller/UdpServicesController.java [hunk 6]
```diff
@@ -53,6 +54,9 @@ public UdpServicesController(OdeProperties odeProps) {
       // PSM internal port
       rm.submit(new PsmReceiver(odeProps));
```
Comment (dan-du-car) [inline jpo-ode-svcs/src/main/java/us/dot/its/jpo/ode/udp/controller/UdpServicesController.java:55]: If we use this GenericReceiver, do we need to submit the above specialized receivers?

File: jpo-ode-svcs/src/main/java/us/dot/its/jpo/ode/udp/map/MapReceiver.java [hunk 7]
```diff
@@ -54,34 +42,20 @@ public void run() {
...
-                    mapMetadata.setRecordGeneratedBy(GeneratedBy.RSU);
-                    mapMetadata.setSecurityResultCode(SecurityResultCode.success);
-
-                    // Submit JSON to the OdeRawEncodedMessageJson Kafka Topic
-                    mapPublisher.publish(JsonUtils.toJson(new OdeAsn1Data(mapMetadata, mapPayload), false),
-                        mapPublisher.getOdeProperties().getKafkaTopicOdeRawEncodedMAPJson());
+                    
+                    String mapJson = UdpHexDecoder.buildJsonMapFromPacket(packet);
+                    if(mapJson != null){
+                        // Submit JSON to the OdeRawEncodedMessageJson Kafka Topic
```
Comment (dan-du-car) [inline jpo-ode-svcs/src/main/java/us/dot/its/jpo/ode/udp/map/MapReceiver.java:48]: This comment is not needed.

File: jpo-ode-svcs/src/main/java/us/dot/its/jpo/ode/udp/psm/PsmReceiver.java [hunk 8]
```diff
@@ -54,34 +42,17 @@ public void run() {
...
-                    psmMetadata.setRecordType(RecordType.psmTx);
-                    psmMetadata.setRecordGeneratedBy(GeneratedBy.UNKNOWN);
-                    psmMetadata.setSecurityResultCode(SecurityResultCode.success);
-
-                    // Submit JSON to the OdeRawEncodedMessageJson Kafka Topic
-                    psmPublisher.publish(JsonUtils.toJson(new OdeAsn1Data(psmMetadata, psmPayload), false),
-                        psmPublisher.getOdeProperties().getKafkaTopicOdeRawEncodedPSMJson());
+                    String psmJson = UdpHexDecoder.buildJsonPsmFromPacket(packet);
+                    if(psmJson != null){
+                        // Submit JSON to the OdeRawEncodedMessageJson Kafka Topic
```
Comment (dan-du-car) [inline jpo-ode-svcs/src/main/java/us/dot/its/jpo/ode/udp/psm/PsmReceiver.java:47]: Same comment as above

File: jpo-ode-svcs/src/main/java/us/dot/its/jpo/ode/udp/spat/SpatReceiver.java [hunk 9]
```diff
@@ -54,34 +42,17 @@ public void run() {
...
-                    spatMetadata.setRecordType(RecordType.spatTx);
-                    spatMetadata.setRecordGeneratedBy(GeneratedBy.RSU);
-                    spatMetadata.setSecurityResultCode(SecurityResultCode.success);
-
+                    
+                    String spatJson = UdpHexDecoder.buildJsonSpatFromPacket(packet);
                     // Submit JSON to the OdeRawEncodedMessageJson Kafka Topic
-                    spatPublisher.publish(JsonUtils.toJson(new OdeAsn1Data(spatMetadata, spatPayload), false),
-                        spatPublisher.getOdeProperties().getKafkaTopicOdeRawEncodedSPATJson());
+                    spatPublisher.publish(spatJson,spatPublisher.getOdeProperties().getKafkaTopicOdeRawEncodedSPATJson());
```
Comment (dan-du-car) [inline jpo-ode-svcs/src/main/java/us/dot/its/jpo/ode/udp/spat/SpatReceiver.java:48]: Is there a reason we do NOT check spat json nullable here, but do check in other receivers.

File: jpo-ode-svcs/src/main/java/us/dot/its/jpo/ode/udp/ssm/SsmReceiver.java [hunk 10]
```diff
@@ -54,34 +42,20 @@ public void run() {
...
-                    ssmMetadata.setRecordGeneratedBy(GeneratedBy.RSU);
-                    ssmMetadata.setSecurityResultCode(SecurityResultCode.success);
-
-                    // Submit JSON to the OdeRawEncodedMessageJson Kafka Topic
-                    ssmPublisher.publish(JsonUtils.toJson(new OdeAsn1Data(ssmMetadata, ssmPayload), false),
-                        ssmPublisher.getOdeProperties().getKafkaTopicOdeRawEncodedSSMJson());
+                    
+                    String ssmJson = UdpHexDecoder.buildJsonSsmFromPacket(packet);
+
+                    if(ssmJson!=null){
```
Comment (dan-du-car) [inline jpo-ode-svcs/src/main/java/us/dot/its/jpo/ode/udp/ssm/SsmReceiver.java:48]: Let us be consistent with this if null check

File: jpo-ode-svcs/src/main/java/us/dot/its/jpo/ode/udp/tim/TimReceiver.java [hunk 11]
```diff
@@ -52,33 +41,19 @@ public void run() {
...
-               timMetadata.setRecordType(RecordType.timMsg);
-               timMetadata.setRecordGeneratedBy(GeneratedBy.RSU);
-               timMetadata.setSecurityResultCode(SecurityResultCode.success);
-
-               // Submit JSON to the OdeRawEncodedMessageJson Kafka Topic
-               timPublisher.publish(JsonUtils.toJson(new OdeAsn1Data(timMetadata, timPayload), false),
-                  timPublisher.getOdeProperties().getKafkaTopicOdeRawEncodedTIMJson());
+               
+               String timJson = UdpHexDecoder.buildJsonTimFromPacket(packet);
+               if(timJson != null){
```
Comment (dan-du-car) [inline jpo-ode-svcs/src/main/java/us/dot/its/jpo/ode/udp/tim/TimReceiver.java:46]: same comment above

File: jpo-ode-svcs/src/main/java/us/dot/its/jpo/ode/coder/stream/LogFileToAsn1CodecPublisher.java [hunk 12]
```diff
@@ -167,9 +167,15 @@ private void publishList(XmlUtils xmlUtils, List<OdeData> dataList) throws JsonP
...
 						publisher.getOdeProperties().getKafkaTopicOdeRawEncodedMAPJson());
+				} else if(messageType == "SPAT"){
+					publisher.publish(JsonUtils.toJson(odeData, false),
+						publisher.getOdeProperties().getKafkaTopicOdeRawEncodedSPATJson());
 				} else if (messageType == "TIM") {
 					publisher.publish(JsonUtils.toJson(odeData, false),
 						publisher.getOdeProperties().getKafkaTopicOdeRawEncodedTIMJson());
+				} else if (messageType == "BSM") {
+						publisher.publish(JsonUtils.toJson(odeData, false),
+							publisher.getOdeProperties().getKafkaTopicOdeBsmJson());
```
[the remark this answers is not in the document]
Comment (John-Wiens) [reply jpo-ode-svcs/src/main/java/us/dot/its/jpo/ode/coder/stream/LogFileToAsn1CodecPublisher.java]: Good Catch - Fixed!

File: jpo-ode-svcs/src/main/java/us/dot/its/jpo/ode/udp/UdpHexDecoder.java [hunk 13]
```diff
@@ -0,0 +1,245 @@
...
+import us.dot.its.jpo.ode.model.OdeSsmMetadata.SsmSource;
+import us.dot.its.jpo.ode.model.OdeTimMetadata;
+import us.dot.its.jpo.ode.model.ReceivedMessageDetails;
+import us.dot.its.jpo.ode.model.RxSource;
+import us.dot.its.jpo.ode.uper.UperUtil;
+import us.dot.its.jpo.ode.util.JsonUtils;
+
+public class UdpHexDecoder {
+    
+    private static Logger logger = LoggerFactory.getLogger(AbstractUdpReceiverPublisher.class);
```
[the remark this answers is not in the document]
Comment (John-Wiens) [reply jpo-ode-svcs/src/main/java/us/dot/its/jpo/ode/udp/UdpHexDecoder.java]: Fixed.

File: jpo-ode-svcs/src/main/java/us/dot/its/jpo/ode/udp/UdpHexDecoder.java (hunk 3 above)
Comment (John-Wiens) [reply jpo-ode-svcs/src/main/java/us/dot/its/jpo/ode/udp/UdpHexDecoder.java:123]: It looks like there is not a Tim Source field defined in the OdeTimMetadata object. Given that all other message types have this we may want to consider adding it. However, doing so will result in a schema change that may break downstream components (Jackson deserialization doesn't accept extra fields by default). In either case, this PR matches the behavior of the existing logic, so adding that field might be out of scope for this PR.

File: jpo-ode-svcs/src/main/java/us/dot/its/jpo/ode/udp/UdpHexDecoder.java (hunk 4 above)
Comment (John-Wiens) [reply jpo-ode-svcs/src/main/java/us/dot/its/jpo/ode/udp/UdpHexDecoder.java]: Sounds good to me. I added a static method at the bottom that is shared between all the sources for this.

File: jpo-ode-svcs/src/main/java/us/dot/its/jpo/ode/udp/bsm/BsmReceiver.java (hunk 5 above)
Comment (John-Wiens) [reply jpo-ode-svcs/src/main/java/us/dot/its/jpo/ode/udp/bsm/BsmReceiver.java]: Fixed.

File: jpo-ode-svcs/src/main/java/us/dot/its/jpo/ode/udp/map/MapReceiver.java (hunk 7 above)
Comment (John-Wiens) [reply jpo-ode-svcs/src/main/java/us/dot/its/jpo/ode/udp/map/MapReceiver.java]: Fixed.

File: jpo-ode-svcs/src/main/java/us/dot/its/jpo/ode/udp/psm/PsmReceiver.java (hunk 8 above)
Comment (John-Wiens) [reply jpo-ode-svcs/src/main/java/us/dot/its/jpo/ode/udp/psm/PsmReceiver.java]: Fixed.

File: jpo-ode-svcs/src/main/java/us/dot/its/jpo/ode/udp/spat/SpatReceiver.java (hunk 9 above)
Comment (John-Wiens) [reply jpo-ode-svcs/src/main/java/us/dot/its/jpo/ode/udp/spat/SpatReceiver.java]: I added a null check.

File: jpo-ode-svcs/src/main/java/us/dot/its/jpo/ode/udp/ssm/SsmReceiver.java (hunk 10 above)
Comment (John-Wiens) [reply jpo-ode-svcs/src/main/java/us/dot/its/jpo/ode/udp/ssm/SsmReceiver.java:48]: I believe the ssmJson has the null check as expected. I didn't make any updates to the specified codeblock. If other changes are required please let me know.

File: jpo-ode-svcs/src/main/java/us/dot/its/jpo/ode/udp/tim/TimReceiver.java (hunk 11 above)
Comment (John-Wiens) [reply jpo-ode-svcs/src/main/java/us/dot/its/jpo/ode/udp/tim/TimReceiver.java:46]: Removed comments, and verified null check.

File: jpo-ode-svcs/src/main/java/us/dot/its/jpo/ode/udp/controller/UdpServicesController.java (hunk 6 above)
Comment (John-Wiens) [reply jpo-ode-svcs/src/main/java/us/dot/its/jpo/ode/udp/controller/UdpServicesController.java:55]: Yes, the generic receiver is in addition to the existing receivers. At this point, I would recommend that we use the specific receivers whenever possible. The Generic Receiver should only be used in deployments where we are unable to separate which messages get sent where. For example, Cohda MK5 RSU units are only able to port forward all messages to a single port right now. Using the specific receivers will be slightly more performant, and will help prevent possible errors caused by message misidentification. Also removing the existing receivers would be a breaking change for nearly all of our deployments.

**Review by dan-du-car — APPROVED** [review]

**Comment by dan-du-car** [comment]
@John-Wiens Is it possible to resolve the conflicts?

**Comment by John-Wiens** [comment]
Fixed Merge Conflicts

**Pull request merged**